src/background.ts

```diff
@@ -1,127 +1,195 @@
-  if (tabs_in_countdown.hasOwnProperty(tabId)) {
+      // tab has same domain as privous update.
+      if (entry.tabIds.includes(tabId)) return;
+    }
+    // move tab to correct domain.
+    // todo: speed this up with extra data struct storing tabId -> domain
+    for (const [originalHostname, blockedEntry] of this._blockedHostnames) {
+      if (originalHostname === key) blockedEntry.tabIds.push(tabId);
+      else {
+        const index = blockedEntry.tabIds.indexOf(tabId);
+        if (index > -1) {
+          blockedEntry.tabIds.splice(index, 1);
+        }
+      }
+    }
+  }
+  isTabBlocked(tabId: number) {
+    for (const [, blockedEntry] of this._blockedHostnames) {
+      const index = blockedEntry.tabIds.indexOf(tabId);
+      if (index > -1) {
+        return blockedEntry.countdown === null || blockedEntry.countdown <= 0;
+      }
+    }
     return false;
   }
-  var hostname = new URL(url).hostname;
-  for (var suffix of blocked_hostnames) {
-    // skip empty strings
-    if (!suffix) {
-      continue;
+  closeTab(tabId: number) {
+    for (const [, blockedEntry] of this._blockedHostnames) {
+      const index = blockedEntry.tabIds.indexOf(tabId);
+      if (index > -1) {
+        blockedEntry.tabIds.splice(index, 1);
+      }
     }
-
-    if (hostname.endsWith(suffix)) {
-      // block this record
-      return true;
+  }
+  addTime(tabId: number, sec: number) {
+    for (const [, blockedEntry] of this._blockedHostnames) {
+      const index = blockedEntry.tabIds.indexOf(tabId);
+      if (index > -1) {
+        blockedEntry.countdown = sec;
+        blockedEntry.tabIds.forEach(this._onCountStart);
+      }
     }
   }
-  return false;
 }
 
-chrome.tabs.onUpdated.addListener((tabId, changeInfo, tab) => {
-  if (changeInfo && changeInfo.status && changeInfo.status == "complete") {
-    // check if this is one of the blocked url
+const stopBlocking = (tabId: number) => {
+  chrome.tabs.sendMessage(tabId, "hide");
+};
 
-    if (!tab.id || !tab.url) return;
-    if (is_hostname_blocked(tab.url, tab.id)) {
-      console.log(tabId, changeInfo, tab, "sending overlay html");
-      // here passing the overlay html, don't think it's the best practice
+const startBlocking = (tabId: number) => {
+  chrome.tabs.sendMessage(tabId, "show");
+};
+
+const onCancel: OnCancelCallback = (tabId) => {
+  stopBlocking(tabId);
+};
+const onTimeout: OnTimeoutCallback = (tabId) => {
+  startBlocking(tabId);
+};
 
-      chrome.tabs.executeScript(tab.id, {
-        file: "overlay/overlay.js",
-      });
-      return;
+const onCountStart: OnCountStartCallback = (tabId) => {
+  stopBlocking(tabId);
+};
+
+const onBlockStart: OnBlockStartCallback = (tabId, isBlocked) => {
+  chrome.tabs.executeScript(
+    tabId,
+    {
+      file: "overlay/overlay.js",
+    },
+    () => {
+      if (isBlocked) startBlocking(tabId);
+      else stopBlocking(tabId);
     }
+  );
+};
+
+const core = new Core([], onCancel, onTimeout, onCountStart, onBlockStart);
```

Comment:
It's for decoupling.  So that `Core` does not depends on `chrome` APIs like `chrome.tabs.*`. I meant to split it out into another ts file. Also, it is more testable, if you want to introduce unit test, you can `expect()` which eventHandlers should be called.

Reply:
Maybe I am over-engineering. What do you think?

Reply:
Good thoughts on testing, but can we just mock on chrome?